aggregator/cmd/main.go

```diff
@@ -69,6 +71,24 @@ func aggregatorMain(ctx *cli.Context) error {
 		return err
 	}
 
+	registry := agg.GetRegistry()
+	rpcServer := rpcserver.NewRpcServer(config.AggregatorServerIpPortAddr, agg, logger)
+	if registry != nil {
+		if err = rpcServer.EnableMetrics(registry); err != nil {
+			return err
+		}
+	}
+
+	go rpcServer.Start()
+
+	restServer := restserver.NewRestServer(config.AggregatorRestServerIpPortAddr, agg, logger)
+	if registry != nil {
+		if err = restServer.EnableMetrics(registry); err != nil {
+			return err
+		}
+	}
+	go restServer.Start()
```

Comment:
I agree with separating the RPC and REST server as modules, but not sure if it makes sense totally separating them from the aggregator, just seems like extra boilerplate. Imo we should still have an `Aggregator::startRpcServer` and `Aggregator::startRestServer` that is called on the construtor or, if that's really the intention here, on the cmd main.

Reply:
Not sure there's any boilerplate. Servers just propagate calls lower to an application but they shall be a sperate entity.

They need to handle various connection scenarios which are not related in any sense to `aggregator`. Those are out of the scope of this PR.

Right now it may look like it is some dummy server propagating calls lower but this PR only intents to separate them  and not add extra logic on top

Reply:
As soon as we will try to introduce extra logic on top of servers: limiting requests rate, reconnections or anything else our aggregator will become a frankenstein handling all of that.

Reply:
No, as I said, I agree with separating them as modules, I don't agree that much with not including them as part of the aggregator and having to create them separately, especially when we have to e.g. expose the prometheus registry from the aggregator in order to do that. I do think they should be separate, I just think the aggregator should have them inside of it and at least some option to create them on the constructor.

Reply:
We don't really have a use-case that justifies this boilerplate above, we always want to expose the aggregator.

Reply:
Well, we do have an issue with that in terms of packages though, since the naive approach would result in a cyclic dependency I think. I don't have an issue keeping the servers in the `aggregator` package though.

Reply:
@emlautarom1 let us know your opinion on this later. Thinking more about it, I don't mind it too much, but I'd still like more opinions.

Reply:
Aggregator shall not know anything about servers. That's IMO incorrect hierarchy. Aggregator is an app for servers. As you noticed having composition or simply creating servers within `Aggregator` will result in circular dependency which is a signal that architecture is broken. 

Having a package that will be only  creating/controlling servers&aggregator is an overkill.

> expose the prometheus registry from the aggregator in order to do that.

That is a hack, I agree but only because of in our eigensdk version API that creates registry itself instead of taking it as an argument.

Reply:
I made a [PR](https://github.com/Layr-Labs/eigensdk-go/pull/117/files) a while ago to eigensdk to improve such behavior. The changes are merged in the later versions.

I guess I can mark this section as TODO for the future removal once we update eigensdk version

Reply:
> Thought of having a package that will be only creating/controlling servers&aggregator but considered it an overkill.

That is my understanding, I agree with the package hierarchy, I just don't like that we have to explicitly expose it through instantiating separate clients when we know we're going to do it everytime. At the same time, with a separate package to handle both we get into a 'subpackage hell'. Would like some others' opinions on that.
My impression is that in go projects people are usually less strict on separating packages.

Reply:
There's an easier solution for this. We can also put the common interfaces in `aggregator/types` - with that, the e.g. RPC package only knows the interface types and the aggregator can just instantiate it, either containing it or just having a utility method for that.

Reply:
I'm not following why servers shall be started from `Aggregator`. 
Such approach basically says - when Aggregator instance is created, we start some servers. That is shall not be true at least for integration/unit tests of aggregator. 

Aggregator - aggregates signatures pretty much and does some db work. Its not clear for me why it shall start servers or anything.

Reply:
Right now we don't have subpackage hell, and I don't think this will lead to this.

The solution that could be applied here, not that I like it. We move servers pack to aggregator package but keep them as separate structures. Create a separate function in utils for creating all of them and conclude on that.

But I don't like above since it is to break heirarchy since compiler won't hit us with circular dependency anymore.

Everything else is attempt to go around compiler in order to make it work. If we need such hacks that means our structure is broken

Reply:
The ideal way would be to have an extra module that holds everything together, but we did agree on it being overkill. That said, separately initializing everything that isn't aggregating would mean separately initializing lots of components of aggregator. I don't think it makes sense it's necessary to add boilerplate for initializing servers all the time when it's something we always want to do - especially if we think about lifetime and such, throwing separate components around is just getting things more complicated than they need to be.

Reply:
IMO making the bridge with aggregator types is the best solution we have here - the servers only know the aggregator interface they need to know and the aggregator can know the servers.

Reply:
Not that I don't agree with the current approach in a way. I'd like more opinions on this - if this takes too long we can also just leave it the way it is and move on.

Reply:
For now, though, one thing I'd ask you to do - since we're going to do the `GetRegistry` hack anyway, you could move the metrics enabling to the constructor itself, it gets considerably less ugly then.

Reply:
After #211 we no longer require to ask the aggregator for the `prometheus.Registry` since we can technically pass it as dependency: currently this is not done but it's a single change to a constructor parameter since the hard work has already been done.

Regarding the question of starting or not the servers in the aggregator: if the aggregator does not need a handle to the servers explicitly then it should **not** be responsible from instantiating them. I don't have enough context on the communication between the aggregator and the servers but if the only thing that they share is this registry then on main we should initialize it, and pass it to the three objects: `aggregator`, `restServer` and `rpcServer`. The less things the aggregator do the better

Reply:
Yeah, as mentioned today, I don't have a problem with it right now, it's fine. What bothers me with initializing the servers outside is lifetime handling - we're passing an aggregator that could be closed

Reply:
> What mainly keeps bothering me with initializing the servers outside is lifetime handling - we're passing an aggregator that could be closed

That's a good point. `Aggregator` only gets closed through a context yet both servers are kept alive due to `http.ListenAndServe`. Ideally, all three services should be tied together with a single context but we cannot easily do that due to `http.ListenAndServe` not accepting a context (see: https://github.com/golang/go/issues/52805).

We need to implement our own graceful shutdown similar to what is described in this SO question: https://stackoverflow.com/questions/39320025/how-to-stop-http-listenandserve. This is outside the scope of this PR but it's worth revisiting.

Reply:
Aggregator doesn't shutdown servers anyway rn. Shall be done on context.Doen similar to what @emlautarom1 linked [here](https://stackoverflow.com/questions/39320025/how-to-stop-http-listenandserve)

Reply:
Resolving this. Will create another PR on `registry` and server shutdown matter